File: app/src/pfp/java/social/entourage/android/privateCircle/PrivateCircleChooseAdapter.java
package social.entourage.android.privateCircle;

import android.content.Context;
import android.graphics.Typeface;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import social.entourage.android.R;
import social.entourage.android.api.model.map.Entourage;

/**
 * Created by Mihai Ionescu on 05/06/2018.
 */
public class PrivateCircleChooseAdapter extends RecyclerView.Adapter {

    private static class PrivateCircleChooseViewHolder extends RecyclerView.ViewHolder {

        protected ImageView privateCircleIcon;
        protected TextView privateCircleTitle;
        protected CheckBox privateCircleCheckbox;

        public PrivateCircleChooseViewHolder(final View itemView, OnCheckedChangeListener onCheckedChangeListener) {
            super(itemView);

            privateCircleIcon = itemView.findViewById(R.id.privatecircle_icon);
            privateCircleTitle = itemView.findViewById(R.id.privatecircle_title);
            privateCircleCheckbox = itemView.findViewById(R.id.privatecircle_checkbox);

            if (privateCircleCheckbox != null) {
                privateCircleCheckbox.setOnCheckedChangeListener(onCheckedChangeListener);
            }

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(final View v) {
                    if (privateCircleCheckbox != null) {
                        privateCircleCheckbox.setChecked(!privateCircleCheckbox.isChecked());
                    }
                }
            });
        }

    }

    private List<Entourage> privateCircleList = new ArrayList<>();
    private int selectedPrivateCircle = AdapterView.INVALID_POSITION;
    private OnCheckedChangeListener onCheckedChangeListener = new OnCheckedChangeListener();

    public void addPrivateCircleList(final List<Entourage> privateCircleList) {
        this.privateCircleList.addAll(privateCircleList);
        selectedPrivateCircle = 0;
        notifyDataSetChanged();
    }

    public int getSelectedPrivateCircle() {
        return selectedPrivateCircle;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_privatecircle_choose_item, parent, false);
        return new PrivateCircleChooseViewHolder(view, onCheckedChangeListener);
    }

    @Override
    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, final int position) {
        Entourage entourage = getItemAt(position);
        if (entourage != null) {
            PrivateCircleChooseViewHolder viewHolder = (PrivateCircleChooseViewHolder)holder;
            Context context = viewHolder.itemView.getContext();

            viewHolder.privateCircleIcon.setImageDrawable(entourage.getIconDrawable(context));

            if (position == selectedPrivateCircle) {
                viewHolder.privateCircleTitle.setTypeface(viewHolder.privateCircleTitle.getTypeface(), Typeface.BOLD);
            } else {
                viewHolder.privateCircleTitle.setTypeface(Typeface.create(viewHolder.privateCircleTitle.getTypeface(), Typeface.NORMAL));
            }
            viewHolder.privateCircleTitle.setText(entourage.getTitle());

            // set the tag to null so that oncheckedchangelistener exits when populating the view
            viewHolder.privateCircleCheckbox.setTag(null);
            // set the check state
            viewHolder.privateCircleCheckbox.setChecked(position == selectedPrivateCircle);
            // set the tag to the item position
            viewHolder.privateCircleCheckbox.setTag(position);
        }
    }

    @Override
    public int getItemCount() {
        return privateCircleList.size();
    }

    protected Entourage getItemAt(final int position) {
        if (privateCircleList == null || position < 0 || position >= privateCircleList.size()) {
            return null;
        }
        return privateCircleList.get(position);
    }

    /**
     * Listener for checkboxes. At most one checkbox can be checked.
     */
    private class OnCheckedChangeListener implements CompoundButton.OnCheckedChangeListener {
        @Override
        public void onCheckedChanged(final CompoundButton compoundButton, final boolean isChecked) {
            // if no tag, exit
            if (compoundButton.getTag() == null) {
                return;
            }
            // get the position
            int position = (Integer) compoundButton.getTag();
            if (position == selectedPrivateCircle) {
                selectedPrivateCircle = AdapterView.INVALID_POSITION;
            } else {
                selectedPrivateCircle = position;
            }
            PrivateCircleChooseAdapter.this.notifyDataSetChanged();
        }
    }
}

File: app/src/pfp/java/social/entourage/android/privateCircle/PrivateCircleChooseAdapter.kt
package social.entourage.android.privateCircle

import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import social.entourage.android.R
import social.entourage.android.api.model.map.Entourage
import java.util.*

/**
 * Created by Mihai Ionescu on 05/06/2018.
 */
class PrivateCircleChooseAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private class PrivateCircleChooseViewHolder(itemView: View, onCheckedChangeListener: OnCheckedChangeListener?) : RecyclerView.ViewHolder(itemView) {
        var privateCircleIcon: ImageView
        var privateCircleTitle: TextView
        var privateCircleCheckbox: CheckBox?

        init {
            privateCircleIcon = itemView.findViewById(R.id.privatecircle_icon)
            privateCircleTitle = itemView.findViewById(R.id.privatecircle_title)
            privateCircleCheckbox = itemView.findViewById(R.id.privatecircle_checkbox)
            if (privateCircleCheckbox != null) {
                privateCircleCheckbox!!.setOnCheckedChangeListener(onCheckedChangeListener)
            }
            itemView.setOnClickListener {
                if (privateCircleCheckbox != null) {
                    privateCircleCheckbox!!.isChecked = !privateCircleCheckbox!!.isChecked
                }
            }
        }
    }

    private val privateCircleList: MutableList<Entourage>? = ArrayList()
    var selectedPrivateCircle = AdapterView.INVALID_POSITION
        private set
    private val onCheckedChangeListener = OnCheckedChangeListener()
    fun addPrivateCircleList(privateCircleList: List<Entourage>?) {
        this.privateCircleList!!.addAll(privateCircleList!!)
        selectedPrivateCircle = 0
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_privatecircle_choose_item, parent, false)
        return PrivateCircleChooseViewHolder(view, onCheckedChangeListener)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val entourage = getItemAt(position)
        if (entourage != null) {
            val viewHolder = holder as PrivateCircleChooseViewHolder
            val context = viewHolder.itemView.context
            viewHolder.privateCircleIcon.setImageDrawable(entourage.getIconDrawable(context))
            if (position == selectedPrivateCircle) {
                viewHolder.privateCircleTitle.setTypeface(viewHolder.privateCircleTitle.typeface, Typeface.BOLD)
            } else {
                viewHolder.privateCircleTitle.typeface = Typeface.create(viewHolder.privateCircleTitle.typeface, Typeface.NORMAL)
            }
            viewHolder.privateCircleTitle.text = entourage.title

            // set the tag to null so that oncheckedchangelistener exits when populating the view
            viewHolder.privateCircleCheckbox!!.tag = null
            // set the check state
            viewHolder.privateCircleCheckbox!!.isChecked = position == selectedPrivateCircle
            // set the tag to the item position
            viewHolder.privateCircleCheckbox!!.tag = position
        }
    }

    override fun getItemCount(): Int {
        return privateCircleList!!.size
    }

    fun getItemAt(position: Int): Entourage? {
        return if (privateCircleList == null || position < 0 || position >= privateCircleList.size) {
            null
        } else privateCircleList[position]
    }

    /**
     * Listener for checkboxes. At most one checkbox can be checked.
     */
    private inner class OnCheckedChangeListener : CompoundButton.OnCheckedChangeListener {
        override fun onCheckedChanged(compoundButton: CompoundButton, isChecked: Boolean) {
            // if no tag, exit
            if (compoundButton.tag == null) {
                return
            }
            // get the position
            val position = compoundButton.tag as Int
            selectedPrivateCircle = if (position == selectedPrivateCircle) {
                AdapterView.INVALID_POSITION
            } else {
                position
            }
            notifyDataSetChanged()
        }
    }
}